### PR TITLE
Spectrum calculator and diagnostics group

### PR DESCRIPTION
**Pull request by slifer50**
# Description

A spectrum calculator built around the Taylor Green problem

I have implemented a spectrum calculator that works for problems with the same size and points in all directions. This has been implemented to work through a diagnostic group. This will form a baseline which can then be extended to work on anisotropic grids and spectrum calculations on the sphere. I've tested this with the Taylor-Green problem and correctly reproduces the results I've gotten by post-processing the vtks from this case. It's been tested to work on a single rank run and with multiple-ranks without issues. @tapios @smarras79. @kpamnany It would be good if you could have a look at the diagnostics implementation to double check everything.

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Diagnostics/groups.jl [hunk 1]
```diff
@@ -321,3 +323,40 @@ function setup_dump_aux_diagnostics(
...
 end
+
+include("dump_spectra.jl")
+"""
+    setup_dump_spectra_diagnostics(
+        ::ClimateMachineConfigType,
+        interval::String,
+        out_prefix::String;
+        writer = NetCDFWriter(),
+        interpol = nothing,
```
Comment (kpamnany) [inline src/Diagnostics/groups.jl:334]:
```suggestion
        interpol = nothing,
        nor = Inf,
```

File: src/Diagnostics/groups.jl [hunk 2]
```diff
@@ -321,3 +323,40 @@ function setup_dump_aux_diagnostics(
...
+`interval` after the velocity fields have been interpolated, into NetCDF files prefixed by
+`out_prefix`.
+"""
+function setup_dump_spectra_diagnostics(
+    ::ClimateMachineConfigType,
+    interval::String,
+    out_prefix::String;
+    writer = NetCDFWriter(),
+    interpol,
+    nor
```
Comment (kpamnany) [inline src/Diagnostics/groups.jl:348]:
```suggestion
    interpol = nothing,
    nor = Inf,
```

File: src/Diagnostics/groups.jl [hunk 3]
```diff
@@ -321,3 +323,40 @@ function setup_dump_aux_diagnostics(
...
+"""
+function setup_dump_spectra_diagnostics(
+    ::ClimateMachineConfigType,
+    interval::String,
+    out_prefix::String;
+    writer = NetCDFWriter(),
+    interpol,
+    nor 
+)
+
```
Comment (kpamnany) [inline src/Diagnostics/groups.jl:350]:
```suggestion
    @assert !isnothing(interpol)
    @assert isfinite(nor)

```

File: src/Common/Spectra/PowerSpectrum.jl [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        if (r[i,j,l] * dx <= k'[endk]) && (r[i,j,l] * dx > (k'[endk] + k'[endk - 1]) / 2)
+          spectrum[endk] = spectrum[endk] + mu[i,j,l]^2 + mv[i,j,l]^2 + mw[i,j,l]^2
+          contribution[endk] = contribution[endk] + 1
+        end
+      end
+    end
+  end
+  spectrum = spectrum .* 2 .* pi .* k.^2 ./ (contribution .* dx.^3) ./ nor
+  return spectrum,k
+end
```
[the remark this answers is not in the document]
Comment (slifer50) [reply src/Common/Spectra/PowerSpectrum.jl:86]: Well it works as intended with range()

**Comment by slifer50** [comment]
bors try

**Comment by slifer50** [comment]
@kpamnany Can you check if everything you requested has been taken care of? If so it would be good if we could go ahead and merge this.

File: experiments/AtmosLES/taylor-green.jl [hunk 5]
```diff
@@ -125,7 +119,7 @@ function config_greenvortex(
         solver_method = LSRK144NiegemannDiehlBusch,
     )
 
-    _C_smag = FT(C_smag(param_set))
+    _C_smag = FT(FT(0.21))
```
Comment (kpamnany) [inline experiments/AtmosLES/taylor-green.jl:122]: Why this change? We should absolutely use `CLIMAParameters` for this; if there's a problem, let's fix it.

Comment (slifer50) [reply experiments/AtmosLES/taylor-green.jl:122]: I could not seem to get this work out of the box, would be glad if you knew how to resolve this.

Comment (slifer50) [reply experiments/AtmosLES/taylor-green.jl:122]: Let's discuss what needs to be done on with regards to the CI on Slack

File: experiments/AtmosLES/taylor-green.jl [hunk 6]
```diff
@@ -21,7 +20,7 @@ using DocStringExtensions
 using LinearAlgebra
 
 using CLIMAParameters
-using CLIMAParameters.Planet: cp_d, MSLP, grav, LH_v0
+using CLIMAParameters.Planet: R_d, cv_d, cp_d, MSLP, grav, LH_v0
```
Comment (kpamnany) [inline experiments/AtmosLES/taylor-green.jl:23]:
```suggestion
using CLIMAParameters.Planet: R_d, cv_d, cp_d, MSLP, grav, LH_v0
using CLIMAParameters.Atmos.SubgridScale: C_smag
```

Comment (kpamnany) [reply experiments/AtmosLES/taylor-green.jl:23]: All the `import` statements below this are unnecessary and can be removed.

File: experiments/AtmosLES/taylor-green.jl (hunk 5 above)
Comment (kpamnany) [reply experiments/AtmosLES/taylor-green.jl:122]: See suggestion above.

File: src/Common/Spectra/PowerSpectrum.jl [hunk 7]
```diff
@@ -0,0 +1,127 @@
+using FFTW
+"""
+  PowerSpectrum(u,v,w,L,dim,nor)
+  Calculates the Powerspectrum from the 3D velocity fields u,v,w
+  inputs need to be equi-spaced and it the domain is assumed to be the same
+  size in all direction and having the same number of points in all directions
+  L : size domain
+  dim : number of points
+  nor : normalization factor
```
Comment (kpamnany) [inline src/Common/Spectra/PowerSpectrum.jl:9]:
```suggestion
    PowerSpectrum(u, v, w, L, dim, nor)

Calculates the Powerspectrum from the 3D velocity fields `u`, `v`, `w`. Inputs
need to be equi-spaced and the domain is assumed to be the same size and
have the same number of points in all directions.

# Arguments
- L: size domain
- dim: number of points
- nor: normalization factor
```

Comment (kpamnany) [reply src/Common/Spectra/PowerSpectrum.jl:9]: Take a look at [this](https://docs.julialang.org/en/v1/manual/documentation/index.html) in general.

File: experiments/AtmosLES/dycoms.jl [hunk 8]
```diff
@@ -326,7 +326,13 @@ function config_dycoms(FT, N, resolution, xmax, ymax, zmax)
...
         solver_method = LSRK144NiegemannDiehlBusch,
     )
-
+    #=
+    ode_solver = ClimateMachine.MISSolverType(;
+        splitting_type = ClimateMachine.SlowFastSplitting(),
+        mis_method = MIS2,
+        fast_method = LSRK144NiegemannDiehlBusch,
+        nsubsteps = 18,
+    )=#
```
Comment (kpamnany) [inline experiments/AtmosLES/dycoms.jl:335]: Did you intend to commit these changes?

Comment (slifer50) [reply experiments/AtmosLES/dycoms.jl:335]: Ah no I was working on another branch fro those. How did they get in?

**Comment by slifer50** [comment]
@simonbyrne, @LenkaNovak, I think the branch is ready for merging, could you have a look and see if there's anything you'd like changed?

**Comment by slifer50** [comment]
bors try

**Comment by charleskawczynski** [comment]
Can we change the file name to `taylor_green.jl`?

**Comment by slifer50** [comment]
> Can we change the file name to `taylor_green.jl`?

@charleskawczynski, sure, I'll change the file name.

**Comment by slifer50** [comment]
bors try

**Comment by kpamnany** [comment]
Not sure what's happened here -- 97 commits and 220 files changes?
